src/Mvc/Mvc.Core/src/ModelBinding/Binders/DateTimeModelBinder.cs

```diff
@@ -0,0 +1,107 @@
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _supportedStyles = supportedStyles;
+            _logger = loggerFactory.CreateLogger<DateTimeModelBinder>();
+        }
+
+        /// <inheritdoc />
+        public Task BindModelAsync(ModelBindingContext bindingContext)
+        {
+            if (bindingContext == null)
+            {
+                throw new ArgumentNullException(nameof(bindingContext));
+            }
+
+            _logger.AttemptingToBindModel(bindingContext);
+
+            var modelName = bindingContext.ModelName;
+            var valueProviderResult = bindingContext.ValueProvider.GetValue(modelName);
+            if (valueProviderResult == ValueProviderResult.None)
+            {
+                _logger.FoundNoValueInRequest(bindingContext);
+
+                // no entry
+                _logger.DoneAttemptingToBindModel(bindingContext);
+                return Task.CompletedTask;
+            }
+
+            var modelState = bindingContext.ModelState;
+            modelState.SetModelValue(modelName, valueProviderResult);
+
+            var metadata = bindingContext.ModelMetadata;
+            var type = metadata.UnderlyingOrModelType;
+            try
+            {
+
+                var value = valueProviderResult.FirstValue;
+
+                object model;
+                if (string.IsNullOrWhiteSpace(value))
+                {
+                    // Parse() method trims the value (with common DateTimeSyles) then throws if the result is empty.
+                    model = null;
+                }
+                else if (type == typeof(DateTime))
+                {
+                    model = DateTime.Parse(value, valueProviderResult.Culture, _supportedStyles);
+                }
+                else
+                {
+                    // unreachable
+                    throw new NotSupportedException();
+                }
+
+                // When converting value, a null model may indicate a failed conversion for an otherwise required
+                // model (can't set a ValueType to null). This detects if a null model value is acceptable given the
+                // current bindingContext. If not, an error is logged.
+                if (model == null && !metadata.IsReferenceOrNullableType)
+                {
+                    modelState.TryAddModelError(
+                        modelName,
+                        metadata.ModelBindingMessageProvider.ValueMustNotBeNullAccessor(
+                            valueProviderResult.ToString()));
+                }
+                else
+                {
+                    bindingContext.Result = ModelBindingResult.Success(model);
+                }
+            }
+            catch (Exception exception)
+            {
+                // Conversion failed.
+                modelState.TryAddModelError(modelName, exception, metadata);
```

Comment:
Thanks, I had a look at the DateTime parsing code as you suggested, and they throw vanilla `FormatException` every time (with the exception of null values which we separately handle): https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeParse.cs#L5065-L5087